odoo/models.py

```diff
@@ -5317,17 +5313,98 @@ def mapped(self, func):
             return self                 # support for an empty path of fields
         if isinstance(func, str):
             recs = self
-            for name in func.split('.'):
-                recs = recs._fields[name].mapped(recs)
+            for field in self._resolve_field_sequence(func):
+                recs = field.mapped(recs)
             return recs
-        else:
+        elif callable(func):
             return self._mapped_func(func)
+        else:
+            raise TypeError("Expected a function or a dot-separated sequence of field names")
+
+    def _filtered_domain_func(self, domain):
+        """Compile the domain as a function
+
+        :param domain: :ref:`A search domain <reference/orm/domains>`.
+        :return: a function ``f`` such that ``f(record)`` returns True if ``record``
+            satisfies the ``domain``.
+        """
+        if not domain:
+            return expression.TRUE_FUNCTION
+
+        def make_leaf_fn(leaf):
+            key, operator, value = expression.normalize_leaf(leaf)
+            # Optimization: short-circuit operator
+            if operator == "=?" and not value:
+                return expression.TRUE_FUNCTION
+            # Special case: hierarchy operators are not implemented in python
+            elif operator in ("child_of", "parent_of"):
+                if not value:
+                    return expression.FALSE_FUNCTION
+                ids = set(self._search([("id", "in", self.ids), leaf], order="id"))
```

Comment:
You mean for domains whose hierachy operator leaf doesn't need to be evaluated due to logical shortcuts?

ie:

```py
domain = [FALSE_LEAF, ("partner_id", "child_of", partner.ids)]
```

It would be possible but I'm not sure it's worth the trouble. Seems like a corner case

IMO the way to improve this is to finally do the python version of hierarchy operators. I tried this already and the perf improvement was superb, specially for the records with parent_store. However, I had to revert it because of some inconsistencies with the SQL implementation that I wasn't aware of. I could've patched my pythonic implementation to make it work like the SQL version, but I honestly thing the SQL imp is broken and the fix has to be done there first, instead.

For example, the case tested here makes no sense IMO:

https://github.com/odoo/odoo/blob/a8dc624635de9d19a24790cd80bb7c2c663e8893/odoo/addons/base/tests/test_expression.py#L120-L129

There are other cases where the current hierarchy operators yield unexpected results when using relational fields in the left operand. I'll try to add some unit tests to prove this point in another PR.